sdk/identity/src/token_credentials/imds_managed_identity_credentials.rs

```diff
@@ -23,104 +29,60 @@ const MSI_API_VERSION: &str = "2019-08-01";
-        self
-    }
-
-    /// Specifies the application id (client id) associated with a user assigned managed service identity resource that should be used to retrieve the access token.
-    ///
-    /// The values of `object_id` and `msi_res_id` are discarded, as only one id parameter may be set when getting a token.
-    #[must_use]
-    pub fn with_client_id<A>(mut self, client_id: A) -> Self
-    where
-        A: Into<String>,
-    {
-        self.client_id = Some(client_id.into());
-        self.object_id = None;
-        self.msi_res_id = None;
-        self
-    }
-
-    /// Specifies the ARM resource id of the user assigned managed service identity resource that should be used to retrieve the access token.
-    ///
-    /// The values of `object_id` and `client_id` are discarded, as only one id parameter may be set when getting a token.
-    #[must_use]
-    pub fn with_identity<A>(mut self, msi_res_id: A) -> Self
-    where
-        A: Into<String>,
-    {
-        self.msi_res_id = Some(msi_res_id.into());
-        self.object_id = None;
-        self.client_id = None;
-        self
-    }
-
     async fn get_token(&self, scopes: &[&str]) -> azure_core::Result<AccessToken> {
         let resource = scopes_to_resource(scopes)?;
 
-        let msi_endpoint = std::env::var(MSI_ENDPOINT_ENV_KEY)
-            .unwrap_or_else(|_| "http://169.254.169.254/metadata/identity/oauth2/token".to_owned());
-
-        let mut query_items = vec![("api-version", MSI_API_VERSION), ("resource", &resource)];
+        let mut query_items = vec![
+            ("api-version", self.api_version.as_str()),
+            ("resource", resource),
+        ];
 
-        match (
-            self.object_id.as_ref(),
-            self.client_id.as_ref(),
-            self.msi_res_id.as_ref(),
-        ) {
-            (Some(object_id), None, None) => query_items.push(("object_id", object_id)),
-            (None, Some(client_id), None) => query_items.push(("client_id", client_id)),
-            (None, None, Some(msi_res_id)) => query_items.push(("msi_res_id", msi_res_id)),
-            _ => (),
+        match self.id {
+            ImdsId::SystemAssigned => (),
+            ImdsId::ClientId(ref client_id) => query_items.push(("client_id", client_id)),
+            ImdsId::ObjectId(ref object_id) => query_items.push(("object_id", object_id)),
+            ImdsId::MsiResId(ref msi_res_id) => query_items.push(("msi_res_id", msi_res_id)),
         }
 
-        let url = Url::parse_with_params(&msi_endpoint, &query_items).context(
-            ErrorKind::DataConversion,
-            "error parsing url for MSI endpoint",
-        )?;
+        let mut url = self.endpoint.clone();
+        url.query_pairs_mut().extend_pairs(query_items);
 
         let mut req = Request::new(url, Method::Get);
 
         req.insert_header("metadata", "true");
 
-        let msi_secret = std::env::var(MSI_SECRET_ENV_KEY);
+        let msi_secret = std::env::var(&self.secret_env);
```

Comment:
This uses `std::env::var`.  Should this be updated to use env from the options?  (see comment from earlier in this file)

Reply:
No, I think that would be better off in a separate PR that aimed at unit testing `get_token`. `Env` is used in the `Credential` `create` functions so that they could be unit tested.